src/nlp/metric.py

```diff
@@ -104,108 +107,123 @@ def __init__(
-                    "data_dir {data_dir}. Using currently defined version "
-                    "{cur_version}.".format(
-                        other_version=str(other_version),
-                        name=self.name,
-                        data_dir=self._data_dir_root,
-                        cur_version=str(self._version),
+    def _create_cache_file(self, timeout=1) -> Tuple[str, FileLock]:
+        """ Create a new cache file. If the default cache file is used, we generated a new hash. """
+        file_path = os.path.join(self.data_dir, f"{self.experiment_id}-{self.num_process}-{self.process_id}.arrow")
+
+        for i in range(self.max_concurrent_cache_files):
+            filelock = FileLock(file_path + ".lock")
+            try:
+                filelock.acquire(timeout=timeout)
+            except Timeout:
+                # If we have reached the max number of attempts or we are not allow to find a free name (distributed setup)
+                # We raise an error
+                if self.num_process != 1:
+                    raise ValueError(
+                        "Another metric instance is already using the local cache file. "
+                        "Please specify an experiment_id to avoid colision between distributed metric instances."
                     )
+                if i == self.max_concurrent_cache_files - 1:
+                    raise ValueError(
+                        f"Cannot acquire lock, too many metric instance are operating concurrently on this file system."
+                        f"You should set a larger value of max_concurrent_cache_files when creating the metric "
+                        f"(current value is {self.max_concurrent_cache_files})."
+                    )
+                # In other cases (allow to find new file name + not yet at max num of attempts) we can try to sample a new hashing name.
+                file_uuid = str(uuid.uuid4())
+                file_path = os.path.join(
+                    self.data_dir, f"{self.experiment_id}-{file_uuid}-{self.num_process}-{self.process_id}.arrow"
                 )
-                logger.warning(warn_msg)
+            else:
+                break
 
-        os.makedirs(version_data_dir, exist_ok=True)
-        return version_data_dir
+        return file_path, filelock
 
-    def _get_cache_path(self, node_id):
-        return os.path.join(self.data_dir, f"{self.experiment_id}-{self.name}-{node_id}.arrow")
+    def _get_all_cache_files(self, timeout=100) -> Tuple[List[str], List[FileLock]]:
+        """ Get a lock on all the cache files in a distributed setup.
+            We wait for timeout second to let all the distributed node finish their tasks (default is 100 seconds).
+        """
+        if self.num_process == 1:
+            file_paths = [self.cache_file_name]
+        else:
+            file_paths = [
+                os.path.join(self.data_dir, f"{self.experiment_id}-{self.num_process}-{process_id}.arrow")
```

Comment:
What about the files with `file_uuid` like this ?

```python
file_path = os.path.join(
    self.data_dir, f"{self.experiment_id}-{file_uuid}-{self.num_process}-{self.process_id}.arrow"
)
```

Reply:
They are only used for single-process metrics (the previous branches of the if-else).
I'll add a comment.